### ISSUE TITLE
State machine can be broken by exceptions

### ISSUE THREAD

**Issue by glenjamin** [bakkerthehacker/bramqp#37]
I'm not 100% on whether this is a problem, or how best to fix it.

The two options I can think of are making the callback in parseFrameEnd use nextTick, or calling resetFrameState before that callback is fired. The second option here probably leaves you needing to handle continuing to read the buffer in the case where there are unread bytes - although you could just trust that more bytes will be along later and will consume parsing.

Anyway here's the code to trigger the error:
```js
'use strict';

var bramqp = require('./');
var net = require('net');
var async = require('async');

process.on('uncaughtException', function(ex) {
    if (ex.message == 'handleable application error') {
        console.warn("Got an app error, but it's ok!");
        return;
    }
    console.warn("Error: %s", ex.stack);
});

var socket = net.connect({
    port : 5672
});
bramqp.initialize(socket, 'rabbitmq/full/amqp0-9-1.stripped.extended', function(error, handle) {

    handle.openAMQPCommunication('guest', 'guest', true, onConnected);

    function onConnected(err) {
        if (err) throw (err);

        handle.queue.declare(1, 'hello');
        handle.once('queue.declare-ok', function(channel, method, data) {
            handle.queue.declare(1, 'world');

            throw new Error('handleable application error');
        });
    }
});

```

**Comment by glenjamin** [comment]
We do get this in production occasionally when an unexpected error occurs, and it also comes up in the test suite for https://github.com/skybet/amqpea when an assertion fails

**Comment by bakkerthehacker** [comment]
This has been fixed in 45492617 and published in 0.3.2

The frame parser reset has been moved into a finally clause where it will always be run, even when throwing an exception.  If there are still issues with this or anything else, feel free to open another issue.
